### ISSUE TITLE
Regression: legacy shipping method partially broken

### ISSUE THREAD

**Issue by k1sul1** [Pakettikauppa/woo-pakettikauppa#96]
If `Show Pakettikauppa shipping method` is set to `yes`, there's an option to add a Pakettikauppa shipping method. 

![image](https://user-images.githubusercontent.com/2719615/68941995-06ec2800-07b0-11ea-92e4-698618481dc0.png)

Adding the shipping method _seems_ to work normally, but the reality is a bit different. There are no notices, warnings or anything of the sort, but the newly added shipping method is missing the title and edit links, as shown here:

![image](https://user-images.githubusercontent.com/2719615/68942073-3864f380-07b0-11ea-9fba-3e0c41dc70c1.png)

If the shipping methods were added in a previous version of the plugin, they continue to work normally, but they can't be edited. How big of an issue is this? 

My understanding is that this is the "legacy way" of doing things, and shouldn't be used anymore. I spent over an hour debugging the problem but didn't find a cause, and I don't know where to look anymore. Instead of spending even more time trying to fix something that shouldn't be used, I propose the following: 

If `Show Pakettikauppa shipping method` is `yes`, and there are shipping methods using it, we show a notice. Something like 

```
if ($legacy_method_enabled) { add_action('admin_notices', function() { echo "We noticed you're still using the legacy Pakettikauppa shipping method. It has been deprecated and will be removed in a future version of the plugin. Please do X instead"; }) }
```

In addition to that, we hide / remove the option on the settings page, so new users won't accidentally use a feature that shouldn't be used, and is entirely broken for new users. 
 
Doing it now also lets us remove the legacy shipping method in a later version, giving users time to migrate. 

@joosev thoughts?

**Comment by joosev** [comment]
I will check it out.

**Comment by joosev** [comment]
The Woo loads stuff in constructor. The title, description etc is loaded in load() function but Woo does not load those at this time, it only calls for constructor.

You can try with this:

    public function __construct( $instance_id = 0 ) {
      parent::__construct($instance_id);

      $this->method_title = 'Foo';
      $this->title = 'Bar';
    }

And see Foo and Bar in the results.

**Comment by joosev** [comment]
I also noticed that         $this->instance_form_fields = $this->my_instance_form_fields();
 is set, even if there is no instance form fields needed (as in, no pakettikauppa shipping methods). I think it should be:

      if ( $show_pakettikauppa_shipping_method === 'yes' ) {
        $this->supports[] = 'instance-settings';
        $this->supports[] = 'instance-settings-modal';
        $this->supports[] = 'shipping-zones';
        $this->instance_form_fields = $this->my_instance_form_fields();
      }

**Comment by k1sul1** [comment]
Woo doesn't call the constructor (anymore). Instead of adding `'\Woo_Pakettikauppa_Core\Shipping_Method'` to `woocommerce_shipping_methods`, I initialize the class first, and then add the instance to to the shipping methods array. 

```php
<?php
add_action(
  'wp_loaded',
  function() {
    $this->shipping_method_instance = $this->load_shipping_method_class();
    $this->add_shipping_method();
  }
);

protected function load_shipping_method_class() {
  require_once 'class-shipping-method.php';

  $method = new Shipping_Method();
  $method->injectCore($this)->load();

  return $method;
}

public function add_shipping_method() {
  add_filter(
    'woocommerce_shipping_methods',
    function( $methods ) {
      $methods[$this->shippingmethod] = $this->shipping_method_instance;

      return $methods;
    }
  );
}
```

As you can see, I call load manually. I had to do it like this so I could inject the core to the shipping method class. Notice that $instance_id in Shipping_Method class is now always 0 (the default). Could that be the problem? If it is that, I don't have any ideas on how to fix it. I can't inject the core class to the shipping method if WooCommerce controls the init. I tried making it a static property earlier, but that won't work either, as it's shared with whitelabel forks as well...

------

I moved `$this->instance_form_fields = $this->my_instance_form_fields();` because it was causing notices, which in turn break the plugin. When the form fields are always set, no notices.

**Comment by k1sul1** [comment]
If you want to get a fix live today, I'd suggest that we just add the deprecation notice that bugs users as long as they keep using the legacy method. The bug looks worse than it is, because IMO it can be mitigated for almost entirely with a simple block of text explaining the situation & instructions on how to fix.

**Comment by joosev** [comment]
The legacy method does not work in 2.1.0 at all.

**Comment by joosev** [comment]
Whitelabel does not need to have support for this legacy shipping method.

**Comment by joosev** [comment]
instance_id != 0 when using the legacy shipping method

**Comment by k1sul1** [comment]
> The legacy method does not work in 2.1.0 at all.

I tested it last week with a fresh installation of 2.0.22. Created an order, upgraded the plugin, created another order, and I was able to create shipping labels for both and change the order status. I'll test it again.

> Whitelabel does not need to have support for this legacy shipping method.

It's not about whitelabel having support for the legacy shipping method. I can't just strip it out in the whitelabel, I'd have replace/rewrite the entire Shipping_Method class.  It's not a bad idea in the long run but that is not going to happen today because I'm not a timelord :slightly_smiling_face:  

> instance_id != 0 when using the legacy shipping method

Well, what is it then? If it's something static the default can be just changed to that, but I doubt it.

**Comment by joosev** [comment]
The legacy shipping method is not shown in the checkout.

when you create new shipping method to shipping class, woo creates unique instance_id for that shipping method. instance_id > 0 when it is added to shipping zone.

**Comment by joosev** [comment]
Problem seems to be, that y ou can't have core in the static.

So, what I did is: I added to Core method "getCore()" which returns the instance of the core.

Then in the shipping_method I created new function called getCore() that will return \Wc_Pakettikauppa->getCore();

and then I replaced all $this->core instances with $this->getCore()

It seems to work.

**Comment by joosev** [comment]
I don't know why, but if you try to save core from the getCore() function to the class, it will fail. It might be a PHP thingie or a Woo or WP thingie. Who knows.

**Comment by joosev** [comment]
Basically by doing this, then in the whitelabel you can override the getCore() function and use own classname and function to get the Core - I think...

**Comment by k1sul1** [comment]
Choppy video but... Indeed, the shipping methods are not listed in the checkout. Old orders can still be completed though.

https://streamable.com/waiao
https://streamable.com/gpnf7

**Comment by joosev** [comment]
So, it will work I think. Tried with this code:

class A {
 public function foo() {
	return 'Hello world';
}
public function bar() {
	return $this->foo();
	}
}

class B extends A {
	public function foo() {
		return 'Nope';
	}
}

$b = new B();

var_dump($b->bar());

===> end results "Nope" == works as it should...

**Comment by joosev** [comment]
In the core:

    final public static function getCore() {
        return self::$core;
    }

and in the construct:

 self::$core = $this;


And in the shipping_method I have:

    public function getCore() {
      return \Wc_Pakettikauppa::getCore();
    }

----

Using this pattern the "injectCore" is not used in this class at all.

**Comment by joosev** [comment]
Then I can use like:         $settings = $this->getCore()->shipment->get_settings();

**Comment by joosev** [comment]
So then in white label we can just use:

public function getCore() {
  return \Wc_WhiteLabel::getCore();
}

**Comment by joosev** [comment]
![Näyttökuva 2019-11-20 kello 14 37 24](https://user-images.githubusercontent.com/14106091/69239614-4d23fb80-0ba3-11ea-836f-d45938977ea1.png)

**Comment by joosev** [comment]
Then it seems to work like it used to work

**Comment by k1sul1** [comment]
You're talking about a singleton. I've used singletons for a lot of things and 99% of the time it's the wrong thing to do, but since we're dealing with WooCommerce, I don't really care, if it works. 

I tried this out, and I can see the legacy method listed on the checkout page, but I still can't edit them.

![image](https://user-images.githubusercontent.com/2719615/69241268-21a31000-0ba7-11ea-8387-5f639d8e6574.png)

Here's a patch: https://gist.github.com/k1sul1/0257b61dca660fa732dde3fa81449c98

(`git apply gistfile1.txt`)

I guess you also changed how the shipping method class is passed to WooCommerce?
